File: mns-reactor/src/main/java/cz/fb/manaus/reactor/rounding/MatchbookRoundingPlugin.java
package cz.fb.manaus.reactor.rounding;

import cz.fb.manaus.core.model.Price;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Component;

import java.util.OptionalDouble;


@Component
@Profile("matchbook")
public class MatchbookRoundingPlugin implements RoundingPlugin {

    @Override
    public OptionalDouble shift(double price, int stepNum) {
        return OptionalDouble.of(price + stepNum * getStep(price));
    }

    public double getStep(double price) {
        return (price - 1) * 0.02;
    }

    @Override
    public OptionalDouble round(double price) {
        return OptionalDouble.of(Price.round(price));
    }

}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/rounding/MatchbookRoundingPlugin.kt
package cz.fb.manaus.reactor.rounding

import cz.fb.manaus.core.model.Price
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Component


@Component
@Profile("matchbook")
class MatchbookRoundingPlugin : RoundingPlugin {

    override fun shift(price: Double, stepNum: Int): Double {
        return price + stepNum * getStep(price)
    }

    private fun getStep(price: Double): Double {
        return (price - 1) * 0.02
    }

    override fun round(price: Double): Double {
        return Price.round(price)
    }

}
